subsys/usb/device_next/class/usbd_hid.c

```diff
@@ -0,0 +1,721 @@
+			       const struct usb_setup_packet *const setup,
+			       struct net_buf *const buf)
+{
+	struct hid_device_data *const ddata = dev->data;
+	struct usbd_hid_descriptor *const desc = ddata->desc;
+
+	if (setup->wValue != 0 || setup->wLength != 1) {
+		errno = -ENOTSUP;
+		return 0;
+	}
+
+	if (desc->if0.bInterfaceSubClass == 0) {
+		/* The device does not support the boot protocol */
+		errno = -ENOTSUP;
+		return 0;
+	}
+
+	LOG_DBG("Get Protocol: %s", ddata->protocol ? "Report" : "Boot");
+	net_buf_add_u8(buf, ddata->protocol);
+
+	return 0;
+}
+
+static int handle_get_descriptor(const struct device *dev,
+				 const struct usb_setup_packet *const setup,
+				 struct net_buf *const buf)
+{
+	struct hid_device_data *const ddata = dev->data;
+	uint8_t desc_type = USB_GET_DESCRIPTOR_TYPE(setup->wValue);
+	uint8_t desc_idx = USB_GET_DESCRIPTOR_INDEX(setup->wValue);
+	struct usbd_hid_descriptor *const desc = ddata->desc;
+
+	switch (desc_type) {
+	case USB_DESC_HID_REPORT:
+		LOG_DBG("Get descriptor report");
+		net_buf_add_mem(buf, ddata->rdesc, MIN(ddata->rsize, setup->wLength));
+		break;
+	case USB_DESC_HID:
+		LOG_DBG("Get descriptor HID");
+		net_buf_add_mem(buf, &desc->hid, MIN(desc->hid.bLength, setup->wLength));
+		break;
+	case USB_DESC_HID_PHYSICAL:
+		LOG_DBG("Get descriptor physical %u", desc_idx);
+		errno = -ENOTSUP;
+		break;
+	default:
+		errno = -ENOTSUP;
+		break;
+	}
+
+	return 0;
+}
+
+static int usbd_hid_ctd(struct usbd_class_node *const c_nd,
+			const struct usb_setup_packet *const setup,
+			const struct net_buf *const buf)
+{
+	const struct device *dev = c_nd->data->priv;
+	int ret = 0;
+
+	switch (setup->bRequest) {
+	case USB_HID_SET_IDLE:
+		ret = handle_set_idle(dev, setup);
+		break;
+	case USB_HID_SET_REPORT:
+		ret = handle_set_report(dev, setup, buf);
+		break;
+	case USB_HID_SET_PROTOCOL:
+		ret = handle_set_protocol(dev, setup);
+		break;
+	default:
+		errno = -ENOTSUP;
+		break;
+	}
+
+	return ret;
+}
+
+static int usbd_hid_cth(struct usbd_class_node *const c_nd,
+			const struct usb_setup_packet *const setup,
+			struct net_buf *const buf)
+{
+	const struct device *dev = c_nd->data->priv;
+	int ret = 0;
+
+	switch (setup->bRequest) {
+	case USB_HID_GET_IDLE:
+		ret = handle_get_idle(dev, setup, buf);
+		break;
+	case USB_HID_GET_REPORT:
+		ret = handle_get_report(dev, setup, buf);
+		break;
+	case USB_HID_GET_PROTOCOL:
+		ret = handle_get_protocol(dev, setup, buf);
+		break;
+	case USB_SREQ_GET_DESCRIPTOR:
+		ret = handle_get_descriptor(dev, setup, buf);
+		break;
+	default:
+		errno = -ENOTSUP;
+		break;
+	}
+
+	return ret;
+}
+
+static void usbd_hid_enable(struct usbd_class_node *const c_nd)
+{
+	const struct device *dev = c_nd->data->priv;
+	struct hid_device_data *ddata = dev->data;
+	const struct hid_device_ops *const ops = ddata->ops;
+	struct usbd_hid_descriptor *const desc = ddata->desc;
+
+	atomic_set_bit(&ddata->state, HID_DEV_CLASS_ENABLED);
```

Comment:
I'm not sure where to put this, and maybe I just didn't find where this happens, but `ddata->protocol` should be reset to REPORT when the HID interface is configured.

Reply:
Maybe we could also call application's callback informing about the protocol change here?

Reply:
below, ready callback is called. Also host should set the protocol whenever initializing a device.

Reply:
So that application is expected to switch to report protocol on ready callback? Sounds ok too.